[Flaky test] Data Race in ci-kubernetes-unit (TestRoundTripSocks5AndNewConnection)

### Which jobs are flaking?

release-master-blocking

- ci-kubernetes-unit

### Which tests are flaking?

- k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy.TestRoundTripSocks5AndNewConnection

Sub-tests:

- `proxied_with_valid_auth_->_https_(invalid_hostname_+_InsecureSkipVerify)`
- `proxied_with_valid_auth_->_http`


![Screenshot from 2023-01-20 23-22-49](https://user-images.githubusercontent.com/73882557/213772128-3f4ddd43-3c8b-4cba-b60b-7f78bc9c6db0.png)


### Since when has it been flaking?

2023-01-20 09:17:28 +0000 UTC

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#ci-kubernetes-unit

### Reason for failure (if possible)

- TestRoundTripSocks5AndNewConnection/proxied_with_valid_auth_->_https_(invalid_hostname_+_InsecureSkipVerify)

```
{Failed  === RUN   TestRoundTripSocks5AndNewConnection/proxied_with_valid_auth_->_https_(invalid_hostname_+_InsecureSkipVerify)
==================
WARNING: DATA RACE
Read at 0x00c00039ca03 by goroutine 609:
  testing.(*common).logDepth()
      /usr/local/go/src/testing/testing.go:889 +0x4e7
  testing.(*common).log()
      /usr/local/go/src/testing/testing.go:876 +0xa4
  testing.(*common).Logf()
      /usr/local/go/src/testing/testing.go:927 +0x6a
  testing.(*T).Logf()
      <autogenerated>:1 +0x75
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing.(*HTTPProxyHandler).ServeHTTP.func2.1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing/http.go:139 +0xcb
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32

Previous write at 0x00c00039ca03 by goroutine 94:
  testing.tRunner.func1()
      /usr/local/go/src/testing/testing.go:1433 +0x7e4
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 609 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing.(*HTTPProxyHandler).ServeHTTP()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing/http.go:138 +0x11e9
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2947 +0x641
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1991 +0xbe4
  net/http.(*Server).Serve.func3()
      /usr/local/go/src/net/http/server.go:3102 +0x58

Goroutine 94 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1726 +0xa84
  main.main()
      _testmain.go:59 +0x2e9
==================
    testing.go:1319: race detected during execution of test
    --- FAIL: TestRoundTripSocks5AndNewConnection/proxied_with_valid_auth_->_https_(invalid_hostname_+_InsecureSkipVerify) (0.11s)
}
```

- TestRoundTripSocks5AndNewConnection/proxied_with_valid_auth_->_http

```
{Failed  === RUN   TestRoundTripSocks5AndNewConnection/proxied_with_valid_auth_->_http
==================
WARNING: DATA RACE
Read at 0x00c000132523 by goroutine 600:
  testing.(*common).logDepth()
      /usr/local/go/src/testing/testing.go:889 +0x4e7
  testing.(*common).log()
      /usr/local/go/src/testing/testing.go:876 +0xa4
  testing.(*common).Logf()
      /usr/local/go/src/testing/testing.go:927 +0x6a
  testing.(*T).Logf()
      <autogenerated>:1 +0x75
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing.(*HTTPProxyHandler).ServeHTTP.func1.1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing/http.go:134 +0xcb
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32

Previous write at 0x00c000132523 by goroutine 94:
  testing.tRunner.func1()
      /usr/local/go/src/testing/testing.go:1433 +0x7e4
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 600 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing.(*HTTPProxyHandler).ServeHTTP()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net/testing/http.go:133 +0x1031
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2947 +0x641
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1991 +0xbe4
  net/http.(*Server).Serve.func3()
      /usr/local/go/src/net/http/server.go:3102 +0x58

Goroutine 94 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1726 +0xa84
  main.main()
      _testmain.go:59 +0x2e9
==================
    testing.go:1319: race detected during execution of test
    --- FAIL: TestRoundTripSocks5AndNewConnection/proxied_with_valid_auth_->_http (0.02s)
}
```

### Anything else we need to know?


### Relevant SIG(s)

/sig api-machinery
cc @kubernetes/ci-signal

--- a/staging/src/k8s.io/apimachinery/pkg/util/net/testing/http.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/net/testing/http.go
@@ -131,13 +131,13 @@ func (h *HTTPProxyHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) 
 	wg.Add(2)
 
 	go func() {
-		defer h.t.Logf("Server read close, host=%s", req.Host)
 		defer wg.Done()
+		defer h.t.Logf("Server read close, host=%s", req.Host)
 		io.Copy(conn, sconn)
 	}()
 	go func() {
-		defer h.t.Logf("Server write close, host=%s", req.Host)
 		defer wg.Done()
+		defer h.t.Logf("Server write close, host=%s", req.Host)
 		io.Copy(sconn, conn)
 	}()
 

